lookout/style/format/feature_extractor.py

```diff
@@ -235,6 +235,74 @@ def extract_features(self, files: Iterable[File], lines: List[List[int]]=None) \
                  the corresponding `VirtualNode`-s and the parents mapping \
                  or None in case no features were extracted.
         """
+        parsed_files, node_parents, vnode_parents = self._parse_vnodes(files, lines)
+        xy = self._files_to_xy(parsed_files)
+        if xy is None:
+            return None
+        X, y, vnodes_y, vnodes, sibling_indices_list = xy
+        self._log.debug("Features shape: %s", X.shape)
+        self._log.debug("Labels shape: %s", y.shape)
+        if self.return_sibling_indices:
+            return X, y, (vnodes_y, vnodes, vnode_parents, node_parents, sibling_indices_list)
+        return X, y, (vnodes_y, vnodes, vnode_parents, node_parents)
+
+    def select_features(self, X: csr_matrix, y: numpy.ndarray) -> Tuple[csr_matrix, numpy.ndarray]:
+        """
+        Select the most useful features based on sklearn's univariate feature selection.
+
+        :param X: Scipy CSR 2-dimensional matrix of features to select.
+        :param y: Numpy 1-dimensional array of labels.
+        :return: Tuple of a CSR matrix with only the selected features (columns) kept and an \
+                 array of the indices of the kept features for later reapplication.
+        """
+        if self.selected_features is not None:
+            return X, self.selected_features
+        feature_selector = VarianceThreshold()
+        X = feature_selector.fit_transform(X)
+        self.selected_features = feature_selector.get_support(indices=True)
+        if self.select_features_number and self.select_features_number < X.shape[1]:
+            feature_selector = SelectKBest(k=self.select_features_number)
+            X = feature_selector.fit_transform(X, y)
+            if self.selected_features is not None:
+                self.selected_features = self.selected_features[feature_selector.get_support(
+                    indices=True)]
+            else:
+                self.selected_features = feature_selector.get_support(indices=True)
+        self._log.debug("Features shape after selection: %s" % (X.shape,))
+        if self.selected_features is None:
+            self.selected_features = numpy.arange(X.shape[1])
+        self._compute_feature_info()
+        return X, self.selected_features
+
+    @property
+    def composite_class_representations(self) -> List[str]:
```

Comment:
should we place properties right after `__main__`?

Reply:
I guess you mean `__init__`